### PR TITLE
RFC: Python Language Support

### PR DESCRIPTION
**Pull request by petrasovaa**
Moved draft RFC from tracwiki:
https://trac.osgeo.org/grass/wiki/RFC/8_LanguageStandardsSupportPython

### CODE REVIEW

**Comment by nilason** [comment]
I see you are having some struggle with linting.
Maybe you could try pre-commit, see

https://github.com/OSGeo/grass/blob/6b2e3099d67f60e187458c393f0c99d5db57af87/doc/development/submitting/submitting.md#use-pre-commit

from #2797

File: doc/development/rfc/python_language_support.md [hunk 1]
```diff
@@ -0,0 +1,32 @@
+# RFC 8: Python Language Support
+
+Author of the first draft: Nicklas Larsson
+
+Status: draft
+
+## Introduction
+
+The code base of GRASS GIS consists today (Feb. 2021) of predominantly C code (ca 50 %), Python (ca 30 %) and a smaller amount of C++ (ca 5 %). Each of these languages have evolved significantly in the last 10–20 years.
```
Comment (petrasovaa) [inline doc/development/rfc/python_language_support.md:9]:
```suggestion
The code base of GRASS GIS consists as of Feb. 2021 of predominantly C code (ca 50 %), Python (ca 30 %) and a smaller amount of C++ (ca 5 %). Each of these languages have evolved significantly in the last 10–20 years.
```

File: doc/development/rfc/python_language_support.md [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+## Background
+
+Similarly, the integration of Python within GRASS GIS – driving the GUI and to a large extent replacing shell scripts (bash/sh) for scripting – has gone through a challenging period with the transition from Python 2 to 3. This transition can now be considered completed, but this fact need to be formalised and the minimum version of Python 3 support must be stated.
+
+The minimum support for Python will, in contrast, likely change more often, due to each (minor) version’s 5-years life-time. At this date, the oldest still-alive Python version is 3.6, to be retired in December 2021.
+
+## Version Updates
+
+For a new release of a minor GRASS version, the Python minimum version should be raised if the current minimum Python version reaches end of life or there are any important technical reasons.
```
Comment (landam) [inline doc/development/rfc/python_language_support.md:22]:
```suggestion
For a new release of a minor GRASS version, the Python minimum version should be raised in the requirements file if the current minimum Python version reaches end of life or there are any important technical reasons.
```

**Review by landam — APPROVED** [review]

File: doc/development/rfc/python_language_support.md (hunk 2 above)
Comment (neteler) [reply doc/development/rfc/python_language_support.md:22]: I suggest to actually link to (using the relative link):

https://github.com/OSGeo/grass/blob/main/REQUIREMENTS.md#general-requirements

File: doc/development/rfc/python_language_support.md [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+At the same time, as for the minor GRASS version, we raise the minimal Python version for a matching branch in grass-addons.
+
+For example, when Python 3.7 goes end-of-life before the release of 8.0.4, we raise the minimal Python version for 8.0.4 to 3.8. We also raise the minimal Python version for grass8 branch in grass-addons repo to Python 3.8.
+
+### References
+
+* Sunsetting Python 2: <https://www.python.org/doc/sunset-python-2/>
+* Active Python Releases: <https://www.python.org/downloads/>
+* Version list: <https://endoflife.date/python>
```
Comment (nilason) [inline doc/development/rfc/python_language_support.md:32]: Perhaps reference to https://devguide.python.org/versions/ would be better?

File: doc/development/rfc/python_language_support.md (hunk 2 above)
Comment (veroandreo) [reply doc/development/rfc/python_language_support.md:22]: > I suggest to actually link to (using the relative link):
> 
> https://github.com/OSGeo/grass/blob/main/REQUIREMENTS.md#general-requirements

done

File: doc/development/rfc/python_language_support.md (hunk 3 above)
Comment (veroandreo) [reply doc/development/rfc/python_language_support.md:32]: > Perhaps reference to https://devguide.python.org/versions/ would be better?

In replacement of the last link there (<https://endoflife.date/python>) or in addition to it?

Comment (nilason) [reply doc/development/rfc/python_language_support.md:32]: Replacement.

Comment (veroandreo) [inline doc/development/rfc/python_language_support.md:32]:
```suggestion
* Version list: <https://devguide.python.org/versions/>
```

**Comment by nilason** [comment]
@petrasovaa Do you mind if I push a 80-col hard wrap change commit on these three RFC PRs. As it stands they are very difficult to read as pure text-files without soft-wrap.

**Comment by nilason** [comment]
I added 80-col line break to this PR, #2813 and #2815.

File: doc/development/rfc/python_language_support.md [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+be raised in the [requirements](https://github.com/OSGeo/grass/blob/main/REQUIREMENTS.md#general-requirements)
+file if the current minimum Python version reaches end of life or there are any
+important technical reasons.
+
+At the same time, as for the minor GRASS version, we raise the minimal Python
+version for a matching branch in grass-addons.
+
+For example, when Python 3.7 goes end-of-life before the release of 8.0.4, we
+raise the minimal Python version for 8.0.4 to 3.8. We also raise the minimal
+Python version for grass8 branch in grass-addons repo to Python 3.8.
```
Comment (petrasovaa) [inline doc/development/rfc/python_language_support.md:43]:
```suggestion
For example, when Python 3.8 goes end-of-life before the release of 8.4.0, we
raise the minimal Python version for 8.4.0 to 3.8. We also raise the minimal
Python version for grass8 branch in grass-addons repo to Python 3.8.
```

**Review by wenzeslaus — APPROVED** [review]

**Review by neteler — APPROVED** [review]

File: doc/development/rfc/python_language_support.md [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+be raised in the [requirements](https://github.com/OSGeo/grass/blob/main/REQUIREMENTS.md#general-requirements)
+file if the current minimum Python version reaches end of life or there are any
+important technical reasons.
+
+At the same time, as for the minor GRASS version, we raise the minimal Python
+version for a matching branch in grass-addons.
+
+For example, when Python 3.8 goes end-of-life before the release of 8.4.0, we
+raise the minimal Python version for 8.4.0 to 3.8. We also raise the minimal
+Python version for grass8 branch in grass-addons repo to Python 3.8.
```
Comment (wenzeslaus) [inline doc/development/rfc/python_language_support.md:43]:
```suggestion
At the same time, we raise the minimal Python version for a matching branch
in grass-addons.

For example, when Python 3.8 goes end-of-life before the release of 8.4.0, we
raise the minimal Python version for 8.4.0 to 3.9. We also raise the minimal
Python version for grass8 branch in grass-addons repo to Python 3.9.
```

**Review by wenzeslaus — CHANGES_REQUESTED** [review]
1) No need to mention GRASS minor version increase for addons, does not add any info and makes the sentence difficult to read.

2) Typo in version number in the example. Context of the first paragraph implies that we are not using version after the end of life.

**Review by wenzeslaus — APPROVED** [review]

**Pull request merged**